Core/src/main/java/com/plotsquared/core/generator/HybridPlotWorld.java

```diff
@@ -267,34 +269,64 @@ public void setupSchematics() throws SchematicHandler.UnsupportedFormatException
         int oddshift = (this.ROAD_WIDTH & 1);
 
         SCHEM_Y = schematicStartHeight();
+
+        // plotY and roadY are important to allow plot and/or road schematic "overflow" into each other without causing AIOOB
+        //  exceptions when attempting either to set blocks to, or get block from G_SCH
+        // Default plot schematic start height, normalized to the minimum height schematics are pasted from.
         int plotY = PLOT_HEIGHT - SCHEM_Y;
         int minRoadWall = Settings.Schematics.USE_WALL_IN_ROAD_SCHEM_HEIGHT ? Math.min(ROAD_HEIGHT, WALL_HEIGHT) : ROAD_HEIGHT;
+        // Default road schematic start height, normalized to the minimum height schematics are pasted from.
         int roadY = minRoadWall - SCHEM_Y;
 
-        int worldHeight = getMaxGenHeight() - getMinGenHeight() + 1;
+        int worldGenHeight = getMaxGenHeight() - getMinGenHeight() + 1;
+
+        int maxSchematicHeight = 0;
+        int plotSchemHeight = 0;
 
         // SCHEM_Y should be normalised to the plot "start" height
         if (schematic3 != null) {
-            if (schematic3.getClipboard().getDimensions().getY() == worldHeight) {
-                SCHEM_Y = plotY = 0;
+            if ((plotSchemHeight = maxSchematicHeight = schematic3.getClipboard().getDimensions().getY()) == worldGenHeight) {
+                SCHEM_Y = getMinGenHeight();
+                plotY = 0;
             } else if (!Settings.Schematics.PASTE_ON_TOP) {
-                SCHEM_Y = plotY = getMinBuildHeight() - getMinGenHeight();
+                SCHEM_Y = getMinBuildHeight();
+                plotY = 0;
             }
         }
 
+        int roadSchemHeight;
+
         if (schematic1 != null) {
-            if (schematic1.getClipboard().getDimensions().getY() == worldHeight) {
-                SCHEM_Y = roadY = getMinGenHeight();
-                if (schematic3 != null && schematic3.getClipboard().getDimensions().getY() != worldHeight
-                        && !Settings.Schematics.PASTE_ON_TOP) {
-                    plotY = PLOT_HEIGHT;
+            if ((maxSchematicHeight = Math.max(
+                    (roadSchemHeight = schematic1.getClipboard().getDimensions().getY()),
+                    maxSchematicHeight
+            )) == worldGenHeight) {
```

Comment:
These assignment expressions make the code really hard to read, I'd prefer to extract it.